storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -2331,10 +2331,12 @@ static bool innodb_init(bool init_dd, bool for_apply_log) {
 
   if (init_dd) {
     dict_load_from_spaces_sdi();
-    dict_sys->dynamic_metadata =
-        dd_table_open_on_name(NULL, NULL, "mysql/innodb_dynamic_metadata",
-                              false, DICT_ERR_IGNORE_NONE);
-    dict_persist->table_buffer = UT_NEW_NOKEY(DDTableBuffer());
+    if (dict_sys->dynamic_metadata == nullptr)
+      dict_sys->dynamic_metadata =
+          dd_table_open_on_name(NULL, NULL, "mysql/innodb_dynamic_metadata",
+                                false, DICT_ERR_IGNORE_NONE);
+    if (dict_persist->table_buffer == nullptr)
+      dict_persist->table_buffer = UT_NEW_NOKEY(DDTableBuffer());
     srv_dict_recover_on_restart();
   }
```

Comment:
I see that upstream commit,  it only moves the dd_table_buffer change little earlier. Can you please share the re-opening stack trace? Aslo check how it worked in previous version? was there any close table before reuse?

Reply:
Since now we saving the metadata after the redo apply phase rather updating the in memory objects. Please check if it is interfering with  PXB-2180 fix.

Reply:
@rahulmalik87 the upstream bug is meant to fix an issue while the server is doing crash recovery. While the server is running and generating MLOG_TABLE_DYNAMIC_META the workflow continues the same. Server generates MLOG_TABLE_DYNAMIC_META and convert them to proper MLOG_INSERT/DELETE... at checkpoint. For PXB we need to still ignore MLOG_TABLE_DYNAMIC_META if we are not in the last incremental backup apply.

Reply:
@satya-bodapati :

>  Can you please share the re-opening stack trace? 

```
InnoDB: Assertion failure: dict0dict.cc:1886:table->get_ref_count() == 0
InnoDB: thread 140139586184512InnoDB: We intentionally generate a memory trap.
InnoDB: Submit a detailed bug report to https://jira.percona.com/projects/PXB.
InnoDB: If you get repeated assertion failures or crashes, even
InnoDB: immediately after the mysqld startup, there may be
InnoDB: corruption in the InnoDB tablespace. Please refer to
InnoDB: http://dev.mysql.com/doc/refman/8.0/en/forcing-innodb-recovery.html
InnoDB: about forcing recovery.
12:58:40 UTC - mysqld got signal 6 ;
Most likely, you have hit a bug, but this error can also be caused by malfunctioning hardware.
Thread pointer: 0x0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0 thread_stack 0x46000
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(my_print_stacktrace(unsigned char const*, unsigned long)+0x55) [0x55fe144a9846]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(handle_fatal_signal+0x2a8) [0x55fe12bd3945]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0) [0x7f74cae323c0]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb) [0x7f74ca47a18b]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x12b) [0x7f74ca459859]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(ut_dbg_assertion_failed(char const*, char const*, unsigned long)+0x150) [0x55fe13324d8b]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(+0x3b7462d) [0x55fe12e6a62d]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(dict_table_remove_from_cache(dict_table_t*)+0x1d) [0x55fe12e6add2]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(dict_close()+0x1ee) [0x55fe12e72cec]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(srv_shutdown()+0x4f1) [0x55fe1328e5ae]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(+0x3092402) [0x55fe12388402]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(+0x309e916) [0x55fe12394916]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(main+0x7bb) [0x55fe123972a3]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f74ca45b0b3]
/work/pxb/ins/8.0/xtrabackup-test/test/../../bin/xtrabackup(_start+0x2e) [0x55fe12380dae]
```

> Aslo check how it worked in previous version? 

We always close the table at srv_shutdown, but before we never enter the place where we would open it on `srv_start` because `srv_log_file_size_requested != srv_log_file_size` and `srv_n_log_files_found != srv_n_log_files` will both evaluate to false, never entering the palce we would open it, thus having only one reference count to innodb_dynamic_metadata.

Reply:
Fixed